src/v/cluster/types.cc

```diff
@@ -293,24 +318,147 @@ std::ostream& operator<<(
+///
+/// Template parameter T is a value type of the tristate.
+template<class T>
+void write_tristate(iobuf& out, const tristate<T>& t) {
+    if (t.is_disabled()) {
+        serde::write(out, static_cast<int8_t>(-1));
+    } else if (!t.has_value()) {
+        serde::write(out, static_cast<int8_t>(0));
+    } else {
+        serde::write(out, static_cast<int8_t>(1));
+        serde::write(out, t.value());
+    }
+}
+
+/// The wrapper is used to serialize topic_configuraiton using serde
+/// library.
+/// The topic_configuration doesn't have default c-tor so it can't be
+/// directly used with serde. Another benefit of using this wrapper is
+/// that serialization details (versions) are not exposed in the header.
+struct topic_configuration_wrapper {
+    using value_t = topic_configuration_wrapper;
+    static constexpr serde::version_t redpanda_serde_version = 0;
+    static constexpr serde::version_t redpanda_serde_compat_version = 0;
+    std::optional<cluster::topic_configuration> cfg;
+
+    void serde_read(iobuf_parser& in, const serde::header& hdr) {
+        auto ns = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+        auto tp = serde::read_nested<ss::sstring>(in, hdr._bytes_left_limit);
+
+        auto partition_count = serde::read_nested<int32_t>(
+          in, hdr._bytes_left_limit);
+        auto replication_factor = serde::read_nested<int16_t>(
+          in, hdr._bytes_left_limit);
+
+        cfg = cluster::topic_configuration(
+          model::ns(ns), model::topic(tp), partition_count, replication_factor);
+
+        cfg->properties.compression = opt_cast<model::compression>(
+          serde::read_nested<std::optional<uint8_t>>(
+            in, hdr._bytes_left_limit));
+        cfg->properties.cleanup_policy_bitflags
+          = opt_cast<model::cleanup_policy_bitflags>(
+            serde::read_nested<std::optional<uint8_t>>(
+              in, hdr._bytes_left_limit));
+        cfg->properties.compaction_strategy
+          = opt_cast<model::compaction_strategy>(
+            serde::read_nested<std::optional<int8_t>>(
+              in, hdr._bytes_left_limit));
+        cfg->properties.timestamp_type = opt_cast<model::timestamp_type>(
+          serde::read_nested<std::optional<uint8_t>>(
+            in, hdr._bytes_left_limit));
+        cfg->properties.segment_size
+          = serde::read_nested<std::optional<size_t>>(
+            in, hdr._bytes_left_limit);
+        cfg->properties.retention_bytes = read_tristate<size_t>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.retention_duration
+          = read_tristate<std::chrono::milliseconds>(in, hdr._bytes_left_limit);
+        cfg->properties.recovery = serde::read_nested<std::optional<bool>>(
+          in, hdr._bytes_left_limit);
+        cfg->properties.shadow_indexing
+          = opt_cast<cluster::shadow_indexing_mode>(
+            serde::read_nested<std::optional<int8_t>>(
+              in, hdr._bytes_left_limit));
```

Comment:
Judging by [this code](https://github.com/vectorizedio/redpanda/blob/6ddc3bb8cd505700e15de1e1d946b12d54f6c9b7/src/v/serde/serde.h#L265-L289) serde won't skip to the end of the envelope for manually defined `serde_read`. @felixguendling can you comment? Maybe we could do envelope size checks and skipping to the envelope end automatically even in this case?

Reply:
Makes sense from my perspective.

Regarding this PR I would probably create a custom read function for `cluster::shadow_indexing_mode` (instead of `opt_cast`).

Additionally, I have a new PR for serde that allows for ADL with`read_nested` so you can skip the template argument.

From my perspective the code should look like this (no `opt_cast` because only the custom read function for `shadow_indexing_mode` is needed to make it work and no `<std::optional<int8_t>>` template because it works with ADL on the second parameter like this:
```cpp
serde::read_nested(in, cfg->properties.shadow_indexing, hdr._bytes_left_limit);
```

I can make a separate PR for enabling ADL if you need it now. Currently, it's in #2433.

Reply:
I'll make a PR with ADL + auto-skip in an hour.

Reply:
The new PR: #2503